File: Core/Algorithms/MinimumInRotatedSortedArray/MinimumInRotatedSortedArray.java
package GoldmanSach.MinimumInRotatedSortedArray;

public class MinimumInRotatedSortedArray {
    /**
     * Finds the minimum element in a rotated sorted array.
     * Time Complexity: O(log N) - Classic Binary Search efficiency.
     * Space Complexity: O(1) - Iterative approach with constant variables.
     */
    public int findMin(int[] nums) {
        // 1. Validation: Prevent processing invalid inputs
        if (nums == null || nums.length == 0) {
            // throw new IllegalArgumentException("Input array must not be null or empty");
            return -1;
        }

        int minimum = nums[0];
        int left = 0, right = nums.length - 1;

        /*
        * We use 'left <= right' because the loop updates an external 'minimum' variable.
        * This ensures that when the search space narrows to a single element (left == right),
        * that element is processed as 'mid' and checked against our minimum before
        * the pointers cross and the loop terminates.
        */
        while (left <= right) {
            /*
            * 2. Shortcut Check:
            * If nums[left] < nums[right], the current window [left...right]
            * is already fully sorted. The minimum in this window MUST be nums[left].
            */
            if (nums[left] < nums[right]) {
                minimum = Math.min(minimum, nums[left]);
                break;
            }

            // 3. Middle Calculation:
            // Using mid = left + (right - left) / 2 is safer for extremely large arrays.
            int mid = (left + right) / 2;
            minimum = Math.min(minimum, nums[mid]);

            /*
            * 4. Search Direction Logic:
            * We compare mid to left to see which "climb" we are currently on.
            */
            if (nums[mid] >= nums[left]) {
                /*
                * Case A: mid is part of the LEFT sorted portion (larger values).
                * Example: [4, 5, 6, 7, 0, 1, 2] -> mid is 7.
                * The true minimum must be to the right of mid.
                */
                left = mid + 1;
            } else {
                /*
                * Case B: mid is part of the RIGHT sorted portion (smaller values).
                * Example: [6, 7, 0, 1, 2, 4, 5] -> mid is 1.
                * The true minimum could be mid or to its left.
                */
                right = mid - 1;
            }
        }

        return minimum;
    }
}

File: Core/Algorithms/MinimumInRotatedSortedArray/minimum-in-rotated-sorted-array.py
from typing import List

"""
Time Complexity: O(n)
Space Complexity: O(1)
"""


def findMinBrute(nums: List[int]) -> int:
    if not nums:
        return None

    minimum = nums[0]
    for i in range(1, len(nums)):
        if nums[i] < minimum:
            minimum = nums[i]
    return minimum


def findMin_(nums):
    return min(nums) if nums else None


"""
Time Complexity: O(logn)
Space Complexity: O(1)
"""


def findMin(nums: List[int]) -> int:
    minimum = nums[0]  # arbitrary value
    l, r = 0, len(nums) - 1

    while l <= r:
        # when array is already linearly sorted i.e. [1,2,3,4,5,6]
        # just return the minimum value
        if nums[l] < nums[r]:
            minimum = min(minimum, nums[l])
            break

        # here, the array is not linearly sorted but rotated
        m = (l + r) // 2
        minimum = min(minimum, nums[m])

        # we want to know if the mid value is a part of the LSP or RSP
        if nums[m] >= nums[l]:
            # search at the right sorted portion
            l = m + 1
        else:
            # search at the left sorted portion
            r = m - 1
    return minimum


def findMinTest(input: tuple):
    value, nums = input

    if not value:
        return nums

    minimum = nums[0]  # or 0

    l, r = 0, len(nums) - 1

    while l <= r:
        if nums[l] < nums[r]:
            minimum = min(nums[l], minimum)
            break

        m = (l + r) // 2
        minimum = min(nums[m], minimum)

        if nums[m] >= nums[l]:
            l = m + 1
        else:
            r = m - 1
    return minimum


# print(f"1: {findMin([4, 5, 6, 7, 0, 1, 2])}")
# print(f"2: {findMin([0, 1, 2, 4, 5, 6, 7])}")
# print(findMinBrute([3, 4, 5, 1, 2]))

testInputs = [(1, [3, 4, 5, 1, 2]), (None, []), (1, [2, 4, 3])]

for input in testInputs:
    print(findMinTest(input))
